File: Algorithms_Part_1/Algorithm1Sample/Ex3WeightedQuickUnion.java
/* *****************************************************************************
 *  Name:
 *  Date:
 *  Description:
 **************************************************************************** */

import edu.princeton.cs.algs4.StdIn;
import edu.princeton.cs.algs4.StdOut;

public class Ex3WeightedQuickUnion {
    private int[] id;
    private int[] size;

    public Ex3WeightedQuickUnion(int n) {
        id = new int[n];
        size = new int[n];
        for (int i = 0; i < n; i++) {
            id[i] = i;
            size[i] = 1;
        }
    }

    private int root(int p) {
        while (p != id[p]) p = id[p];
        return p;
    }

    public void union(int p, int q) {
        int pid = root(p);
        int qid = root(q);
        if (pid == qid) return;
        if (size[pid] >= size[qid]) {
            id[pid] = qid;
            size[pid] += size[qid];
        }
        else {
            id[qid] = pid;
            size[qid] += size[pid];
        }
    }

    public boolean isConnected(int p, int q) {
        return root(p) == root(q);
    }

    @Override
    public String toString() {
        StringBuilder result = new StringBuilder();
        for (int i = 0; i < id.length; i++)
            result.append(id[i]).append(" ");
        result.append("size:");
        for (int i = 0; i < id.length; i++)
            result.append(size[i]).append(" ");
        return result.toString();
    }

    public static void main(String[] args) {

        int n = StdIn.readInt();
        Ex3WeightedQuickUnion qf = new Ex3WeightedQuickUnion(n);
        while (!StdIn.isEmpty()) {
            int p = StdIn.readInt();
            int q = StdIn.readInt();

            if (!qf.isConnected(p, q)) {
                qf.union(p, q);
                StdOut.println("Connected(" + p + "," + q + ") : " + qf.toString());
            }
        }

    }
}

File: PythonLearnig/Algorithms_Part_1/Ex3WeightedQuickUnion.py
class WeightedQuickUnion:

    def __init__(self, n):
        self.data = list(range(n))
        self.size = [1] * n

    def union(self, p, q):
        pid = self.root(p)
        qid = self.root(q)

        if self.size[pid] >= self.size[qid]:
            self.data[pid] = qid
            self.size[pid] += self.size[qid]
        else:
            self.data[qid] = pid
            self.size[qid] += self.size[pid]

    def is_connected(self, p, q):
        return self.root(p) == self.root(q)

    def __str__(self):
        return str(self.data)

    def root(self, p):
        while p != self.data[p]:
            p = self.data[p]
        return p


def test_weighted_quick_union():

    n = int(input("Enter the no of nodes: "))
    m = int(input("Enter the no of unions: "))

    wqu = WeightedQuickUnion(n)
    for i in range(m):
        p = int(input())
        q = int(input())

        if not wqu.is_connected(p, q):
            wqu.union(p, q)
            print(f"Connected ({p},{q}) : {wqu}")


if __name__ == '__main__':
    test_weighted_quick_union()
